Clinton david/mcdropout

clinton0313 opened the pull request:
Initial pull request to start reviewing code for eventual adoption.

Commit:
Allow mulitplication of discrete and continuous spaces

Commit:
Revert "Allow mulitplication of discrete and continuous spaces"

This reverts commit ecd930c493cb1f5902d387e8cb4f4d8703a54318.

clinton0313 committed:
Added a skeleton framework for MCDropout and Dropconnect. Untested

clinton0313 committed:
Started  builder function. Untested.

clinton0313 committed:
Added a few more comments and type hints for next steps.

uri-granta committed:
Support ExpectedFeasibility with LocalPenalization (#516)

uri-granta committed:
try_get_optimal_point helper method (#517)

uri-granta committed:
Test with numpy typing stubs (#521)

uri-granta committed:
Define mulitplication on mixed spaces and equality between spaces (#518)

henrymoss committed:
Inducing point selectors (#511)

clinton0313 committed:
Added test_dense_forward and some test files

clinton0313 committed:
p_dropout test implemented. Fixed some import issues. Deleted MCDropConnect Layer.

clinton0313 committed:
p_dropout test fixed. fit test implemented. Removed relative import dots

clinton0313 committed:
Started builder test

uri-granta committed:
More TensorBoard monitoring improvements (#515)

clinton0313 committed:
Finished test_builders.py cannot try it due to relative importing issues

uri-granta committed:
Update to gpflow 2.4.0 (#522)

clinton0313 committed:
Moved test files

uri-granta committed:
Fix quickdocs inconsistency and make it error if it happens again (#525)

clinton0313 committed:
Debugged DropoutNetwork and DropconectNetwork in architectures

Debugged build_vanilla_keras_mcdropout

test_builders_dropout passed.

clinton0313 committed:
Removed mcdropout tester notebook

clinton0313 committed:
Added docstrings to dropout architectures and layers.

clinton0313 committed:
Added test_architectures_dropout.py

-test model attributes
-test can be compiled

clinton0313 committed:
In architectures, layers, builders changed naming of parameter dropout_prob to rate to be consistent with keras Dropout

In architectures: Fixed DropConnectNetwork's output layer dropout rate.

Comleted test_architectures_dropout with all tests passing.

clinton0313 committed:
Tightened variance of dropout rate in dropout rate test.

clinton0313 committed:
Merge pull request #2 from clinton0313/clinton/tests

Important fixes to architectures.

Changed parameter of dropout rate from dropout_prob to rate to be consistent with Keras Dropout.

Added testing for architectures.

uri-granta committed:
Use GPFlow posterior objects for great good (#523)

ampudia19 committed:
basic model and notebook

clinton0313 committed:
Added dropout test to architectures test

Changed dropout rate checks to only include floats between 0 and 1 to be consistent with keras Dropout layer in:
        layers.py
        architectures.py

Updated rate parameter and typing hint in dropout builder test.

ampudia19 committed:
fixed naming and dropout layer

clinton0313 committed:
Merge pull request #3 from clinton0313/clinton/integration_tests

Added dropout test to architectures test

ampudia19 committed:
add model sample function

ampudia19 committed:
Merge branch 'clinton_david/mcdropout' of https://github.com/clinton0313/trieste into clinton_david/mcdropout

henrymoss committed:
Compiled model optimizers for gpflow models (#528)

henrymoss committed:
New TS notebook (#527)

henrymoss committed:
Raise error if use SVGP model with q_diag=True (#533)

clinton0313 committed:
Updated gitingore.

hstojic commented on docs/notebooks/deep_mcdropout.py:
```diff
@@ -0,0 +1,106 @@
+
```
this is good, but lacks the BO part to include it in trieste

hstojic replied:
I guess you plan to do that part later on?

hstojic commented on trieste/__init__.py:
```diff
@@ -25,6 +25,10 @@
 
 .. bibliography::
 """
+
+import os, sys; sys.path.append(os.path.dirname(os.path.realpath(__file__)))
+
+
```
what is this part about?

clinton0313 replied:
This was back when I was struggling with the relative imports and trying all sorts of stuff. Missed removing this.

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -19,11 +19,13 @@
 from __future__ import annotations
 
 from abc import abstractmethod
+from multiprocessing.sharedctypes import Value
```
is this really necessary?

hstojic replied:
I actually don't see it used anywhere?

clinton0313 replied:
This was an accidental automatic import, typing in VS code.

hstojic commented on trieste/models/keras/builders.py:
```diff
@@ -80,3 +81,40 @@ def build_vanilla_keras_ensemble(
+    units: int = 50,
+    activation: str | tf.keras.layers.Activation = "relu",
+    rate: Sequence[float | int] | float | int = 0.5,
+    dropout: str = "standard"
```
what about passing the network, `DropoutNetwork` or `DropConnectNetwork` instead of the string which gives additional space for errors?
you can still put `DropoutNetwork` as a default...

hstojic commented on trieste/models/keras/builders.py:
```diff
@@ -80,3 +81,40 @@ def build_vanilla_keras_ensemble(
+    num_hidden_layers: int = 3,
+    units: int = 50,
+    activation: str | tf.keras.layers.Activation = "relu",
+    rate: Sequence[float | int] | float | int = 0.5,
```
these are important defaults, ones we think work well in BO context - might need revision after some experience with using MCdropout in BO loop

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,115 @@ def optimize(self, dataset: Dataset) -> None:
+        :param num_samples: The number of samples at each point.
+        :return: The samples, with shape [..., S, N].
+        """
+        return tf.stack([self.model(query_points, training=True) for _ in range(num_samples)], axis=0)
```
why training=True?

hstojic replied:
now its clearer, but worth explaining this in the docstring above, or the docstring of the class

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,115 @@ def optimize(self, dataset: Dataset) -> None:
+        :return: The predicted mean and variance of the observations at the specified
+            ``query_points``.
+        """
+        T = 100
```
avoid hard coding, this should be an argument to the model

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,115 @@ def optimize(self, dataset: Dataset) -> None:
         self.model.fit(x=x, y=y, **self.optimizer.fit_args)
+
+class MCDropout(KerasPredictor, TrainableProbabilisticModel):
+    def __init__(
```
missing a class docstring

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,115 @@ def optimize(self, dataset: Dataset) -> None:
+        model: DropoutNetwork,
+        optimizer: Optional[KerasOptimizer] = None
+    ) -> None:
+
```
missing init docstring

hstojic replied:
missing init docstring

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,115 @@ def optimize(self, dataset: Dataset) -> None:
         x, y = self.prepare_dataset(dataset)
         self.model.fit(x=x, y=y, **self.optimizer.fit_args)
+
+class MCDropout(KerasPredictor, TrainableProbabilisticModel):
```
you should be able to implement a trajectory sampler for this model easily, then you can add `HasTrajectorySampler` class, that would allow you to use continuous thompson sampling acquisition function with this model

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
         return input_tensor, output_tensor
+
+
+class DropoutNetwork(KerasEnsembleNetwork):
```
KerasEnsembleNetwork was meant to define a standard for elements of ensemble type of model, so it doesn't make sense that you inherit from this class, even if it works in practice - you are creating a class like `KerasEnsemble`, do you really need any method from `KerasEnsembleNetwork`? I don't see why any of them would be necessary or useful here?

hstojic replied:
actually, what you probably should do is subclassing `tf.keras.Model`, check option 2 in https://www.tensorflow.org/api_docs/python/tf/keras/Model

I found it a bit difficult to do that for the `KerasEnsemble` (I still plan to go back to it at some point), but here it should be straightforward, so please try to do that

clinton0313 replied:
Ok, if I am doing this, is it still necessary to explicitly indicate the input shape for Trieste?

clinton0313 replied:
I am struggling a bit to understand what is needed for Trieste. It seems that if subclassing from tf.keras.Model all that is needed is to create a model class that implements the correct Call method with  the number of hidden layers and output layer. In that case the builder method would only need to  use the data's output dimension to build the network and keras can infer the input shape when fit is first called? Is that right?

hstojic replied:
what is needed by trieste  are methods defined by interfaces, how exactly a model looks like is more up to specific model wrapper, e.g. in DeepEnsemble model argument needs to be KerasEnsemble (not `tf.keras.Model`) and then that wrapper can make sure the model is correctly initialized etc.

In case of subclassing from tf.keras.Model after creating an instance of the model, calling the object with some data will create the network, I think you don't need to call fit. This could be done in the init method of the model wrapper.

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+        self.rate = rate
+        self._model = self._build_model()
+
+    @property
+    def model(self) -> tf.keras.Model:
+        return self._model
```
if you subclass tf.keras.Model, then self is instance of it and there is no need for this property

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+            self._check_probability(rate)
+            self._rate = [rate for _ in range(len(self._hidden_layer_args) + 1)]
+        else:
+            raise TypeError(f"dropout_prob needs to be a sequence, float or int. Instead got {type(rate)}")
```
I don't see a condition for `int`?

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+                    Got {len(rate)} dropout probabilities for {len(self._hidden_layer_args) + 1} layers."""
+                )
+            for p in rate:
+                self._check_probability(p)
```
check `tf.debugging` you should be able to do this check with functions there, no need for this `_check_probability` thing

clinton0313 replied:
I tried to take a look at the tf.debugging module, but I couldn't find anything of the sort? There are regular assert greater than or equal to's. I also looked at the tensorflow Dropout layer and there they just use a similar raise ValueError check.

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+    def rate(self) -> Sequence[float]:
+        return self._rate
+    
+    @rate.setter
```
is rate meant to be changed after the model is instantiated?

clinton0313 replied:
Probably not. I was more aiming for giving the flexibility of setting the rate with either a single float or a list of floats, if for some reason it was desirable to pass different probabilities for example. 

The setting of rates has been used as convenience in some of the tests which could be written to avoid it.

hstojic replied:
but that would still be passed when the object/network is created, no? it seems to me that regardless of whether you give a single float or a list you wouldn't change it afterwards so this setter is not needed

if you don't see a clear need for this additional flexibility its better to simplify, if really needed this can be amended later on

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,148 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+    def _gen_hidden_layers(self, input_tensor: tf.Tensor) -> tf.Tensor:
+
+        for index, hidden_layer_args in enumerate(self._hidden_layer_args):
+            layer_name = f"{self.network_name}dense_{index}"
```
i needed to set names carefully bcs dealing with ensemble of models is a bit more involved (in terms of assigning inputs/outputs correctly), so you can dispense with the custom naming

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -19,11 +19,13 @@
 import numpy as np
 import tensorflow as tf
 import tensorflow_probability as tfp
+from .layers import DropConnect, MCDropoutLayer
```
MCDropoutLayer is not used anywhere?

hstojic commented on trieste/models/keras/layers.py:
```diff
@@ -0,0 +1,76 @@
+        return outputs
+
+        
+class MCDropoutLayer(Dropout):
+    def call(self, x, **kwargs):
+        kargs = {k:v for k,v in kwargs.items() if k != "training"}
+        return super().call(x, training=True, **kargs)
```
what is this exactly doing?

hstojic commented on trieste/models/keras/layers.py:
```diff
@@ -0,0 +1,76 @@
+
+
+class DropConnect(Dense):
+    def __init__(self, rate: Union[float, int]=0.5, *args, **kwargs):
```
why `int`?

hstojic commented on trieste/models/keras/layers.py:
```diff
@@ -0,0 +1,76 @@
+        "param **kwargs: Keyword arguments passed to Dense Keras class
+        """
+        self.rate = rate
+        super(DropConnect, self).__init__(*args, **kwargs)
```
I think you don't need the `DropConnect, self` part

hstojic commented on trieste/models/keras/layers.py:
```diff
@@ -0,0 +1,76 @@
+
+    @rate.setter
+    def rate(self, rate):
+        assert 0 <= rate <= 1, f"prob needs to be a valid probability instead got {rate}"
```
you should avoid using asserts in code, they are meant for tests, use tf.debugging for checking things

hstojic reviewed (commented):
overall looks like its in a decent shape, left comments throughout, lets start working through them

- re drop connect can you give me the reference article? I would perhaps leave that out from trieste at this stage, let me give it a think 

I didn't go through the tests carefully in this round, superficially they are mostly ok - the only comments there is that I think you are missing tests that would trigger those checks (say invalid rates), and test_model is missing

hstojic commented:
by the way, in pyproject file at the root you have various python formatting and typing commands that you need to run and probably make some changes after seeing the output (black, isort, flake8, mypy)

clinton0313 committed:
Changed test_predictions and models to temporarily include passes as a parameters

clinton0313 committed:
Removed layer names. Removed unecessary imports. Fixed typing of rate parameters. Removed MCDropoutLayer.

clinton0313 committed:
Started remaking architecture to inherit tf.keras.Model

clinton0313 committed:
Used tf dropout function in Dropconnect to make it cleaner and more accurate. Amended test_layers to not use rate of 1.

clinton0313 committed:
Remade MCdropout architecture

clinton0313 committed:
Completed redone architecure inheriting from Tensorflow Model and updated architecture tests for dropout.

clinton0313 committed:
Amended builders and builder tests to work with new architecture.

clinton0313 committed:
Removed unecessary import

clinton0313 committed:
Updated architecture

clinton0313 committed:
Merge branch 'clinton/pr_fixes' into clinton/integration_tests

Remade DropoutNetwork architecture to inherit from Keras Model Class.

Adjusted builders, test_builders, test_architectures to work with new architecture

Adjusted DropConnect Layer to more accurate way of dropping out.

Adjusted test_layer to work with new DropConnect Layer

Cleaned unecessary imports in init files.

apaleyes committed:
HIPPO: HIghly Parallelizable Pareto Optimization (#519)

uri-granta committed:
Fix slowtests broken by gpflow posterior pr (#532)

hstojic committed:
sgpr wrapper (#531)

* sgpr wrapper draft

* wip

* fixed tests

* fixed trajectory sampling; check for num_latent_gp; formattingh

* fixing typing; integration tests

* fixed latent gp checks and tests

* fixed covariance test

* fixed trajectory sampler test

clinton0313 commented:
I ran black and flake8, over the root directory and it actually ended up changing many files outside of the ones we have been working on. Should I include all of this when I merge the branch I have been working on or leave those out?

clinton0313 committed:
Simplified rate parameter.

Adjusted architecture, layers and builders tests to reflect changed rate parameter.

clinton0313 committed:
Adde docstring to DropConnect layer.

clinton0313 committed:
Reformatted documents and cleaned up imports

clinton0313 committed:
Set sensible defaults for DropoutNetwork and DropConnectNetwork and associated wrappers

Implemented integration test for DropoutNetworks

clinton0313 committed:
Removed scripts folder

clinton0313 committed:
Adjusted DropoutNetwork to infer dtype from input data. Added and adjusted relevant unit and integration tests.

clinton0313 committed:
Updated DropOutNetwork and DropConnectNetwork docstrings.

hstojic committed:
added integration test for parallel TS

ampudia19 committed:
sampler, first test

ampudia19 committed:
Merge pull request #4 from clinton0313/clinton/integration_tests

Clinton/integration tests

clinton0313 committed:
Implemented BO tests in separate file that work, as well as deep mc dropout notebook, but there are issues with estimating variance in the model. Unable to pass tests.

clinton0313 committed:
Added upranking in DropoutNetwork architecture.

clinton0313 committed:
Modified notebooks

clinton0313 committed:
Added architecture test to check for scalars

clinton0313 committed:
Quick fix to dropout_network not accepting list of scalars as a tf.keras.Model

clinton0313 committed:
Fixed model predict not computing variance properly issue.

hstojic committed:
working draft, no tests

hstojic committed:
fixing samples option

hstojic committed:
another fix of the sampling mode

ampudia19 committed:
Merge pull request #5 from clinton0313/clinton/dropout_network_quickfix

Quick fix to a shape issue in DropoutNetwork's new model architecture

ampudia19 committed:
Merge branch 'clinton_david/mcdropout' into david/sampler

ampudia19 committed:
Merge branch 'hstojic/de_sampler_update' into david/sampler

clinton0313 committed:
Changed plotting function for plotly to show sequence of predictions.

ampudia19 committed:
fixed sampler

clinton0313 committed:
Removed separate dropout test files and merged to original test files. Marked mcdropout tests.

clinton0313 committed:
Refactored test_layers.py

clinton0313 committed:
Changed plotly colour to black for best point

ampudia19 committed:
final sampling tests, init models tests

ampudia19 committed:
Merge remote-tracking branch 'origin/clinton/BO_tests' into david/sampler

ampudia19 committed:
progress model tests

clinton0313 committed:
Quick and dirty fix to reset learning rates in MCDropout optimize

Updated docstrings and default settings to architectures, builders and model.

ampudia19 committed:
more models tests, LR fix in models

clinton0313 committed:
Moved BO tests to original file. Updated layers. Updated MCDropout notebook.

clinton0313 committed:
Removed separate bayesian optimization pytest file.

ampudia19 committed:
Merge pull request #7 from clinton0313/clinton/BO_tests

Bayesian Optimization integration tests

ampudia19 committed:
Merge branch 'clinton_david/mcdropout' into david_sampler

ampudia19 committed:
cleaning

ampudia19 committed:
Merge branch 'david_sampler' of https://github.com/clinton0313/trieste into david/sampler

clinton0313 committed:
Fixed couple small typos in bayesian optimization tests for mcdropout

hstojic commented on docs/notebooks/util/plotting_plotly.py:
```diff
@@ -51,8 +52,8 @@ def format_point_markers(
     """
-
-    col_pts = np.repeat(c_pass, num_pts).astype("<U15")
+    cm = plt.get_cmap("RdYlGn")
+    col_pts = list(np.arange(num_pts)) #np.repeat(c_pass, num_pts).astype("<U15")
```
what are these changes for?

clinton0313 replied:
These can be left out but it really helped me diagnose the model. It colors the query points in order. Here is an example:

![dropconnect_035_fit](https://user-images.githubusercontent.com/88747458/163407251-72599e6f-b074-4072-8675-f72002b046aa.png)
.

hstojic replied:
ok, fair enough, that does improve the figure - but that would probably need to be mentioned in other notebooks that have a similar figure, what the change in the color means...

clinton0313 replied:
Done in the Deep evidential regression branch which will be eventually merged here.

hstojic commented on docs/notebooks/deep_mcdropout.py:
```diff
@@ -46,61 +52,278 @@ def objective(x, error=True):
+
+
+# %% [markdown]
+# We can visualise how the optimizer performed as a three-dimensional plot. Crosses mark the initial data points while dots mark the points chosen during the Bayesian optimization run. You can see that there are some samples on the flat regions of the space, while most of the points are exploring the ridges, in particular in the vicinity of the minimum point.
```
has the text been updated to reflect the results you obtain here?

clinton0313 replied:
No, the notebook has yet to be updated and finished with text and comments.

hstojic commented on docs/notebooks/deep_mcdropout.py:
```diff
@@ -46,61 +52,278 @@ def objective(x, error=True):
-        rate=rate,
-        dropout=dropout
+
+# def build_cubic_model(data: Dataset) -> MCDropout:
```
what do you get here, something reasonable?

clinton0313 replied:
Yes, the cubic model seemed reasonable and fine. The notebook is a bit of a mess because I was more merging it in for David to pull and then work on.

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -423,6 +506,32 @@ def _test_optimizer_finds_minimum(
+                rate=0.1
+            dropout_network = build_vanilla_keras_mcdropout(initial_data,  rate=rate, dropout_network=DropoutNetwork)
+        
+        fit_args = {
+            "batch_size": 32,
+            "epochs": 1000,
+            "callbacks": [
+                tf.keras.callbacks.EarlyStopping(monitor="loss", patience=80), 
+                tf.keras.callbacks.ReduceLROnPlateau(monitor="loss", factor=0.3, patience=15)
+                ],
+            "verbose": 0
+        }
```
this might be a sensible default to use in the wrapper? if thats the case then we can set it there and here these lines can be removed

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -344,6 +352,81 @@ def test_bayesian_optimizer_with_deep_ensemble_finds_minima_of_simple_quadratic(
+        pytest.param(0.25, 5, EfficientGlobalOptimization(), id="EfficientGlobalOptimization"),
+        pytest.param(0.1, 5, DiscreteThompsonSampling(500, 1), id="DiscreteThompsonSampling")
+    ],
+)
```
please add parallel thompson sampling as well, once the trajectory sampling is up for the task...

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -344,6 +352,81 @@ def test_bayesian_optimizer_with_deep_ensemble_finds_minima_of_simple_quadratic(
 
+@random_seed
+@pytest.mark.slow
+@pytest.mark.mcdropout
```
what is this bit?? defined in conftest?

clinton0313 replied:
I marked all of mcdropout tests so it was convenient to run pytest tests/unit/models/keras -m "mcdropout" for example to run all of mcdropout specific tests quickly.

hstojic replied:
ah, thats nice, wasnt aware you could do that, can you please mark the ones specific to deep ensembles with it own flag?

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -344,6 +352,81 @@ def test_bayesian_optimizer_with_deep_ensemble_finds_minima_of_simple_quadratic(
+@pytest.mark.parametrize(
+    "rate, num_steps, acquisition_rule",
+    [
+        pytest.param(0.25, 5, EfficientGlobalOptimization(), id="EfficientGlobalOptimization"),
```
if you have rate here why do you have rule specific rates below in the model setup part?

clinton0313 replied:
These were missed changes that I have submitted to change in a small PR.

ampudia19 replied:
Clinton, did you run the tests after your changes? I am getting errors running it with your latest code,and it's solved by adding rate as an argument to the pytest functions

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -344,6 +352,81 @@ def test_bayesian_optimizer_with_deep_ensemble_finds_minima_of_simple_quadratic(
+
+
+@random_seed
+@pytest.mark.slow
```
we have simple quadratic tests because they should be fast, so these shouldnt be marked slow...

hstojic commented on tests/integration/test_bayesian_optimization.py:
```diff
@@ -344,6 +352,81 @@ def test_bayesian_optimizer_with_deep_ensemble_finds_minima_of_simple_quadratic(
 
 
+@random_seed
+@pytest.mark.slow
```
we have simple quadratic tests because they should be fast, so these shouldnt be marked slow...

hstojic commented on tests/scripts/basic_dropout.ipynb:
```diff
@@ -0,0 +1,284 @@
+{
```
this file should be removed no?

clinton0313 replied:
Yes. I thought it was already removed.

hstojic commented on trieste/models/keras/builders.py:
```diff
@@ -82,39 +81,51 @@ def build_vanilla_keras_ensemble(
     data: Dataset,
-    num_hidden_layers: int = 3,
-    units: int = 50,
+    num_hidden_layers: int = 5,
+    units: int = 300,
```
this seems a bit large, no?

clinton0313 replied:
This was just what seemed to work well when doing the integration test to see how well it could learn the function with a certain dropout rate. More testing might need to be done to find parameters that might be smaller, but it needs to be balanced with dropout rate. 

A smaller network has to have a smaller dropout rate to be able to still learn the function. A smaller dropout rate seems to sometimes lead to underexploration which then makes it slower even though the network took longer to train. 

For example: 

with EI using a rate of 0.1 and the above network size, seems like it has started moving towards the optimum but has slowed down due to the valley and is exploring too slowly potentially. :
![dropout_EI_01_fit](https://user-images.githubusercontent.com/88747458/163408904-e72ac83b-cdb1-45a4-ad84-af1a8af07121.png)

Then using a rate of 0.2:
![dropout_EI_02_fit](https://user-images.githubusercontent.com/88747458/163408963-5241038c-e8b6-4fa7-9939-6ea70d28140e.png)

Then using a rate of 0.25 and unable to learn the function:
![dropout_EI_025_fit](https://user-images.githubusercontent.com/88747458/163409034-9334e78d-d4e0-4a3c-9342-ee7087148265.png)

But yes, I think more testing needs to be done to find better defaults potentially and optimize it to the best speed. I kind of thought that would be part of benchmarking and trying different parameters.

hstojic replied:
with integration test you mean BO integration test or learning the function, i.e. prediction test?

clinton0313 replied:
I meant prediction test. I did try a few different parameters in the BO test as well and that was what seemed to work well for now, but potentially need to do more testing to find better defaults.

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -333,11 +333,13 @@ def optimize(self, dataset: Dataset) -> None:
         model: DropoutNetwork,
-        optimizer: Optional[KerasOptimizer] = None
+        optimizer: Optional[KerasOptimizer] = None,
+        num_passes: int = 100,
```
perhaps `num_samples` would be a better name?

ampudia19 replied:
`num_samples` is an argument in the function `sample`, which is different to the number of forward passes used to compute the mean and variance. I'd keep it as is, but I'll change it upon confirmation.

hstojic replied:
fair enough, lets leave it as is... think about a good default value as well

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -333,11 +333,13 @@ def optimize(self, dataset: Dataset) -> None:
         self.model.fit(x=x, y=y, **self.optimizer.fit_args)
 
+
 class MCDropout(KerasPredictor, TrainableProbabilisticModel):
```
lets call it `MonteCarloDropout`, a bit clearer and still not too long

ampudia19 replied:
Sounds good.

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -386,8 +395,9 @@ def optimize(self, dataset:Dataset) -> None:
         :param dataset: The data with which to optimize the model.
         """
         x, y = dataset.astuple()
+        self.optimizer.optimizer.learning_rate = self._learning_rate
```
its probably safer to reset it after calling `fit`

this is also probably wrong way to do it,you should use `assign` method :
```
self.optimizer.optimizer.learning_rate.assign(self._learning_rate)
```

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -237,146 +237,167 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
-            {"units": 50, "activation": "relu"},
-            {"units": 50, "activation": "relu"},
-            {"units": 50, "activation": "relu"},
+            {"units": 300, "activation": "relu"},
+            {"units": 300, "activation": "relu"},
+            {"units": 300, "activation": "relu"},
+            {"units": 300, "activation": "relu"},
+            {"units": 300, "activation": "relu"}
```
this seems like an overkill for small datasets common in BO, no?

hstojic commented on trieste/models/keras/architectures.py:
```diff
@@ -237,146 +237,167 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+        output_layer = tf.keras.Sequential(name="output_layer")
+        output_layer.add(tf.keras.layers.Dropout(
+            rate=self.rate, 
+            dtype=self.input_tensor_spec.dtype
```
setting the dtype here is sufficient, no need for input layer?
have you tested it with both float64 and float32 input?

clinton0313 replied:
I just double checked, and I made a test for this. It is titled test_dropout_network_dtype in test_architectures.py file. I believe it is a sufficient test?

clinton0313 replied:
I just double checked, and I made a test for this. It is titled test_dropout_network_dtype in test_architectures.py file. I believe it is a sufficient test?

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +332,129 @@ def optimize(self, dataset: Dataset) -> None:
+        if self.optimizer.loss is None:
+            self.optimizer.loss = "mse"
+
+        self._learning_rate = self.optimizer.optimizer.learning_rate.numpy()
```
yes, this is a good solution, `DeepGaussianProcess` class in gpflux package does it as well...

ampudia19 committed:
predict, loss model tests

hstojic commented on tests/unit/models/keras/test_architectures.py:
```diff
@@ -262,3 +280,142 @@ def test_gaussian_network_is_correctly_constructed(
+    assert dropout_nn.optimizer is None
+
+    # check correct number of layers and proerply constructed
+    assert len(dropout_nn.layers) == 2
```
should the value on the right depend on `num_hidden_layers` argument?

clinton0313 replied:
No because I have a Sequential layer that contains all the hidden layers and then a Sequential layer that contains the output layers. Could be potentially simplified, but I thought it organized it nicely since typically all the hidden layers are the same and the output layer is different.

hstojic reviewed (commented):
looks quite good already, left some comments

ampudia19 committed:
Merge pull request #8 from clinton0313/clinton/BO_tests

Fixed couple small typos in bayesian optimization tests for mcdropout

ampudia19 committed:
last models test

ampudia19 committed:
MCDropout -> MonteCarloDropout

ampudia19 committed:
Merge branch 'clinton_david/mcdropout' into david_sampler

ampudia19 committed:
Merge pull request #6 from clinton0313/david_sampler

Sampler, LR fix and tests for sampler and models

ampudia19 committed:
appended model tests

hstojic commented:
ok, now you'll have to merge my draft deep ensemble sampler PR into `final_mcdropout` :)

hstojic commented on trieste/models/keras/sampler.py:
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+
+        return dropout_trajectory(self._model)
+
+    def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunction:
```
you'll need `update_trajectory` method as well, you can copy the one from the deep ensemble sampler I think

hstojic commented on trieste/models/keras/sampler.py:
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+        trajectory.resample()  # type: ignore
+        return trajectory
+
+# @tf.function
```
this shouldn't be here I think

hstojic commented on trieste/models/keras/sampler.py:
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+            tf.ones([0,0], dtype=tf.int32), shape=[None, None], trainable=False
+        )
+            
+    def __call__(self, x: TensorType) -> TensorType:  # [N, B, d] -> [N, B]
```
you need tf.function decorator for this method

ampudia19 replied:
Yes, that's what I was working on in one of the branches, but failed to find a way of making it deterministic. Discussing it with Clinton, he proposed a viable alternative, but it would require some recoding of the layers and elsewhere in the model. I was going to bring it up tomorrow in our morning meeting.

hstojic commented on trieste/models/keras/sampler.py:
```diff
@@ -80,29 +83,183 @@ def resample_trajectory(self, trajectory: TrajectoryFunction) -> TrajectoryFunct
+        batch_index = tf.range(0, self._batch_size, 1)
+        _ = self._model.sample(x[:1,0,:], num_samples=1) # [DAV] Somehow first seed doesn't propagate unless I've done a dummy pred before.
+        for b, seed in zip(batch_index, tf.unstack(self._seeds)):
+            tf.random.set_seed(seed) # [DAV] A possible local seed? One can pass operational seeds to both types of dropouts, but it doesn't seem to fix the prediction
```
@ampudia19 setting a global TF seed seems wrong way to go about this, its also incomplete, few more seeds need to be set to make sure things are really deterministic...

can we have a method in the model that sets/masks weights in the dropout layers where we generate this with a seed and a rate? then we could call this method for each trajectory and with seeds that you generated here, and this would make sure we are having the same trajectory?

ampudia19 replied:
All deterministic tests I proposed, it passed. But it is clearly incompatible with tf.function, so a better solution is indeed necessary.

Clinton proposed a very similar solution to yours. I'll happily implement it, and will be nice to talk over it tomorrow.

hstojic replied:
there is an oldish discussion here https://stackoverflow.com/questions/56707661/manually-assign-dropout-layer-in-keras that might be useful

hstojic replied:
ah, this seems like exactly what we need: https://github.com/keras-team/keras/issues/14172

hstojic replied:
(although set_weights might be computationally expensive...)

ampudia19 replied:
I tried setting weights with a minimum working example, but I couldn't get_weights while in graph execution..

ampudia19 replied:
Perhaps the behavior of set_weights is completely different from that of get_weights.

hstojic replied:
another option could be to overwrite Dropout layer class and somehow allow for seed argument

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +338,176 @@ def optimize(self, dataset: Dataset) -> None:
+        model: DropoutNetwork,
+        optimizer: Optional[KerasOptimizer] = None,
+        num_passes: int = 100,
+        learning_rate: float = 0.01
```
what is this input? its not declared below as self?

ampudia19 replied:
It is already deleted in other branches, but not pushed. It is a residual argument from when Clinton and I discussed how to be go about resetting LR.

hstojic commented on trieste/models/keras/models.py:
```diff
@@ -332,3 +338,176 @@ def optimize(self, dataset: Dataset) -> None:
+            callback with patience of 50, and verbose 0.
+            See https://keras.io/api/models/model_training_apis/#fit-method for a list of possible
+            arguments.
+        :raise ValueError: If ``model`` is not an instance of
```
missing parameter description

ampudia19 committed:
write up dropout notebook

ampudia19 committed:
new function checks for notebook

clinton0313 committed:
Fixed a few missed MCDroput renames to MonteCarloDropout

ampudia19 committed:
additional function

ampudia19 committed:
fixed init functions

clinton0313 committed:
Removed unecessary learning_rate attribute
